### PR TITLE
feat(behavior_velocity): improve occlusion spot detection area polygon

### PR DESCRIPTION
**Pull request by taikitanaka3**
## Description(required)

image

This PR improves
- detection area polygon shape which can be affect ego behavior(private road only)
- make polygon dynamic for consider ego state
- calculate rough ego ttc from ego velocity and create DA polygon from interpolated distance that pedestrian can reach within ego ttc.

This PR wont fix
- integrate best ego state calculation at time t.https://github.com/autowarefoundation/autoware.universe/issues/383
This can be achieved by having velocity smoother internally or trajectory feedback form velocity smoother or making planning manager or constant velocity of ego ?
- best occlusion spot detection using opencv library.
Maybe some method can help finding out occlusion spot more efficiently?
- projection of dynamic object and lanelet map info to occupancy grid
- smooth slow down limit. https://github.com/autowarefoundation/autoware.universe/pull/388

This PR should be after 
~https://github.com/autowarefoundation/autoware.universe/pull/344 -> merged~

image

## Review Procedure(required)

1. set this parameter in dummy_perception_publiher.launch.xml
2. set Perception/OccupancyGrid/Map tab true in Rviz
`<param name="enable_ray_tracing" value="false" />`
see unit test covered this case
image

run test 
```
1: [----------] 1 test from detectionArea
1: [ RUN      ] detectionArea.calcLateralDistance
1: s: 0 v: 5 d: 2.5
1: s: 5 v: 5 d: 3.5
1: s: 10 v: 5 d: 5
1: s: 15 v: 5 d: 5
1: [       OK ] detectionArea.calcLateralDistance (1 ms)
1: [----------] 1 test from detectionArea (1 ms total)
```

```
2: [----------] 1 test from isOcclusionSpotSquare
2: [ RUN      ] isOcclusionSpotSquare.occlusion_spot_cell
2: TooNoisyCase->No OcclusionSpot 2by2
2:     0|1|2|3|4|                      
2:   0  |?| | | |                      
2:   1  |?| | | |                      
2:   2 ?|?|?|?| |                      
2: [       OK ] isOcclusionSpotSquare.occlusion_spot_cell (1 ms)
2: [----------] 1 test from isOcclusionSpotSquare (1 ms total)
```

3 launch psim and see occlusion spot module works

https://user-images.githubusercontent.com/65527974/153832554-2a95d320-a4df-4c71-ad22-8e2c576f9be1.mp4

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [ ] Read [commit-guidelines][commit-guidelines]
- [ ] Assign PR to reviewer

If you are adding new package following items are required:

- [ ] Documentation with description of the package is available
- [ ] A sample launch file and parameter file are available if the package contains executable nodes

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[commit-guidelines]: https://www.conventionalcommits.org/en/v1.0.0/
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/scene_occlusion_spot_in_private_road.cpp [hunk 1]
```diff
@@ -48,11 +48,14 @@ bool OcclusionSpotInPrivateModule::modifyPathVelocity(
   if (path->points.size() < 2) {
     return true;
   }
-  param_.vehicle_info.baselink_to_front = planner_data_->vehicle_info_.max_longitudinal_offset_m;
-  param_.vehicle_info.vehicle_width = planner_data_->vehicle_info_.vehicle_width_m;
-
+  // set planner data
+  {
+    param_.v.j_max = planner_data_->max_stop_jerk_threshold;
+    param_.v.a_max = planner_data_->max_stop_acceleration_threshold;
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/scene_occlusion_spot_in_private_road.cpp:54]: j/a_max is minimum value. So, should it be j/a_min ?

**Comment by taikitanaka3** [comment]
minor change until 5af18dc
- sidewalk -> detection area
- square poly -> triangle poly

image

**Comment by taikitanaka3** [comment]
@tkimura4 cc: @yukkysaito 
Because of dynamic polygon generation, discrete lat lon slice polygon creation could be rattling.
so changed shape to longitudinal slice polygon and only consider more notable(most closest) occlusion at each slice compare to backward slice.
image

**Comment by taikitanaka3** [comment]
feat(behavior_velocity): better occlusion spot detection improves
from
image
to
image

File: planning/behavior_velocity_planner/config/occlusion_spot.param.yaml [hunk 2]
```diff
@@ -1,6 +1,7 @@
 /**:
   ros__parameters:
     occlusion_spot:
+      debug: true                          # [-] whether to publish debug markers. Note Default should be false for performance
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/config/occlusion_spot.param.yaml:4]: Note need to set false before merge

File: planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp [hunk 3]
```diff
@@ -228,17 +251,17 @@ void calcSlowDownPointsForPossibleCollision(
...
   const LaneletMapPtr lanelet_map_ptr, const double max_range, const PathWithLaneId & path,
   const ROAD_TYPE & target_road_type);
-//!< @brief convert a set of occlusion spots found on sidewalk slice
-void generateSidewalkPossibleCollisionFromOcclusionSpot(
-  std::vector<PossibleCollisionInfo> & possible_collisions, const grid_map::GridMap & grid,
-  const std::vector<grid_map::Position> & occlusion_spot_positions,
-  const double offset_from_start_to_ego, const lanelet::ConstLanelet & path_lanelet,
-  const PlannerParam & param);
+//!< @brief convert a set of occlusion spots found on detection_area slice
+boost::optional<PossibleCollisionInfo> generateOneNotebleCollisionFromOcclusionSpot(
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp:255]: typo: Noteble -> Notable

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/geometry.cpp [hunk 4]
```diff
@@ -16,120 +16,113 @@
...
+  const double min_distance = (is_on_right) ? -p.half_vehicle_width : p.half_vehicle_width;
+  const double slice_length = p.detection_area.slice_length;
+  const int num_step = static_cast<int>(slice_length);
   //! max index is the last index of path point
-  const int max_index = static_cast<int>(inner_bounds.size() - 1);
+  const int max_index = static_cast<int>(center_line.size() - 2);
+  int idx = 0;
+  /**
+   * Note: create polygon from path point is better than from ego offset to consider below
+   * - less computation cost and no need to recalulated interpolated point start from ego
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/geometry.cpp:71]: typo: recalulated -. recalculated

File: planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp (hunk 3 above)
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp]: [f087f9e](https://github.com/autowarefoundation/autoware.universe/pull/354/commits/f087f9e37f13853ee18711199e750fdc747c638c)

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/geometry.cpp (hunk 4 above)
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/src/scene_module/occlusion_spot/geometry.cpp]: [f087f9e](https://github.com/autowarefoundation/autoware.universe/pull/354/commits/f087f9e37f13853ee18711199e750fdc747c638c)

**Comment by yukkysaito** [comment]
memo
- introduce the maximum distance of the detection area
- Need to discuss whether to use a fixed value or use the output of smoother for the lateral distance of the detection area

**Comment by taikitanaka3** [comment]
@yukkysaito 
>introduce the maximum distance of the detection area
this should be with https://github.com/autowarefoundation/autoware.universe/pull/388

**Comment by taikitanaka3** [comment]
@yukkysaito 
> introduce the maximum distance of the detection area
> Need to discuss whether to use a fixed value or use the output of smoother for the lateral distance of the detection area
these changes are major update and will be on another PR

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp [hunk 5]
```diff
@@ -481,10 +466,15 @@ void generateSidewalkPossibleCollisionFromOcclusionSpot(
...
+    // skip if absolute distance is larger
+    if (distance_lower_bound < dist) continue;
     const double length_to_col = arc_coord_occlusion_point.length - baselink_to_front;
     ArcCoordinates arc_coord_collision_point = {
       length_to_col,
       calcSignedLateralDistanceWithOffset(arc_coord_occlusion_point.distance, half_vehicle_width)};
+    // skip if occlusion is behind ego bumper
     if (length_to_col < offset_from_start_to_ego) {
       continue;
     }
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp:498]: I think this judge should be in Line474.

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/scene_occlusion_spot_in_private_road.cpp [hunk 6]
```diff
@@ -82,18 +85,20 @@ bool OcclusionSpotInPrivateModule::modifyPathVelocity(
...
+    detection_area_polygons, interp_path, offset_from_start_to_ego, param_);
   RCLCPP_DEBUG_STREAM_THROTTLE(logger_, *clock_, 3000, "closest_idx : " << closest_idx);
   RCLCPP_DEBUG_STREAM_THROTTLE(
     logger_, *clock_, 3000, "offset_from_start_to_ego : " << offset_from_start_to_ego);
   std::vector<utils::PossibleCollisionInfo> possible_collisions;
   // Note: Don't consider offset from path start to ego here
-  utils::generateSidewalkPossibleCollisions(
-    possible_collisions, grid_map, interp_path, offset_from_start_to_ego, param_,
-    debug_data_.sidewalks);
+  utils::generateDetectionAreaPossibleCollisions(
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/scene_occlusion_spot_in_private_road.cpp:98]: getPossibleCollisionsinDetectionArea?

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp [hunk 7]
```diff
@@ -415,61 +416,45 @@ void filterCollisionByRoadType(
   }
 }
 
-void generateSidewalkPossibleCollisions(
+void generateDetectionAreaPossibleCollisions(
+  std::vector<Slice> & detection_area_polygons,
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp:438]: `const` can add to `detection_area_polygons`

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp (hunk 5 above)
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp:495]: [a6ceb8c](https://github.com/autowarefoundation/autoware.universe/pull/354/commits/a6ceb8c20e4da72920718d7ce4b6ac1e1e2ce1e1)

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/geometry.cpp [hunk 8]
```diff
@@ -16,120 +16,113 @@
...
 {
 using lanelet::BasicLineString2d;
 using lanelet::BasicPoint2d;
 using lanelet::BasicPolygon2d;
 namespace bg = boost::geometry;
 namespace lg = lanelet::geometry;
 
-void createOffsetLineString(
-  const BasicLineString2d & in, const double offset, BasicLineString2d & offset_line_string)
+BasicPoint2d calculateOffsetPoint(
```
[the remark this answers is not in the document]
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/src/scene_module/occlusion_spot/geometry.cpp]: [b067661](https://github.com/autowarefoundation/autoware.universe/pull/354/commits/b0676615e86c71d3083f35e6221c0be97147a3a9)

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp (hunk 7 above)
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp]: [b067661](https://github.com/autowarefoundation/autoware.universe/pull/354/commits/b0676615e86c71d3083f35e6221c0be97147a3a9)

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/scene_occlusion_spot_in_private_road.cpp (hunk 6 above)
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/src/scene_module/occlusion_spot/scene_occlusion_spot_in_private_road.cpp]: I changed to createPossibleCollisionInDetectionArea
[b067661](https://github.com/autowarefoundation/autoware.universe/pull/354/commits/b0676615e86c71d3083f35e6221c0be97147a3a9)

**Comment by taikitanaka3** [comment]
@yukkysaito  @tkimura4 
change was not huge so I add here
[feat(behavior_velocity): introduce detection area max length](https://github.com/autowarefoundation/autoware.universe/pull/354/commits/9cd6a6ad7570674372a9516cb45aaaa879be26e8)

File: planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp [hunk 9]
```diff
@@ -78,6 +78,8 @@ struct Velocity
   double max_stop_jerk;         // [m/s^3] emergency braking system jerk
   double max_stop_accel;        // [m/s^2] emergency braking system deceleration
   double max_slow_down_accel;   // [m/s^2] maximum allowed deceleration
+  double non_effective_jerk;    // [m/s^3] too weak jerk for velocity planning.
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp:81]: `too` is misleading. Simply, this is no problem with `weak jerk for velocity planning.`

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/geometry.cpp [hunk 10]
```diff
@@ -59,7 +59,10 @@ void buildSlices(
    * +--------------------------+ - +---+(max_length,min_distance) = inner_polygons
    */
   const double min_length = offset;  // + p.baselink_to_front;
-  const double max_length = std::min(p.detection_area_max_length, p.detection_area_length);
+  // Note: min_detection_area_length is for occlusion spot visualization but not effective for planning
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/geometry.cpp]: this is currently for visualization before reaching 
image

File: planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp (hunk 9 above)
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp:81]: fixed at 
[bfea98d](https://github.com/autowarefoundation/autoware.universe/pull/354/commits/bfea98d867805564378c02de02bfe78b6ffdbaba)

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/scene_occlusion_spot_in_public_road.cpp [hunk 11]
```diff
@@ -56,6 +56,9 @@ bool OcclusionSpotInPublicModule::modifyPathVelocity(
     param_.v.max_stop_accel = planner_data_->max_stop_acceleration_threshold;
     param_.v.v_ego = planner_data_->current_velocity->twist.linear.x;
     param_.v.a_ego = planner_data_->current_accel.get();
+    param_.detection_area_max_length = planning_utils::calcJudgeLineDistWithJerkLimit(
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/scene_occlusion_spot_in_public_road.cpp:59]: Could you add a description about detection_area_length to the document?

Comment (taikitanaka3) [reply planning/behavior_velocity_planner/src/scene_module/occlusion_spot/scene_occlusion_spot_in_public_road.cpp:59]: updated at 
[1b3e3e3](https://github.com/autowarefoundation/autoware.universe/pull/354/commits/1b3e3e305c351bef940936f066dc53b713d67475)

**Review by yukkysaito — APPROVED** [review]

**Comment by taikitanaka3** [comment]
- [x] done writing to release note

**Pull request merged**